File: src/main/java/com/github/nsc/de/shake/parser/node/expression/AddNode.java
package com.github.nsc.de.shake.parser.node.expression;

import com.github.nsc.de.shake.parser.node.ValuedNode;
import com.github.nsc.de.shake.parser.node.factor.DoubleNode;
import com.github.nsc.de.shake.parser.node.factor.IntegerNode;

public class AddNode extends ExpressionNode {

    public AddNode(ValuedNode left, ValuedNode right, int operatorPosition) {
        super(left, right, operatorPosition);
    }
    public AddNode(int left, ValuedNode right, int operatorPosition) {
        this(new IntegerNode(left), right, operatorPosition);
    }
    public AddNode(ValuedNode left, int right, int operatorPosition) {
        this(left, new IntegerNode(right), operatorPosition);
    }
    public AddNode(int left, int right, int operatorPosition) {
        this(new IntegerNode(left), new IntegerNode(right), operatorPosition);
    }
    public AddNode(double left, ValuedNode right, int operatorPosition) {
        this(new DoubleNode(left), right, operatorPosition);
    }
    public AddNode(ValuedNode left, double right, int operatorPosition) {
        this(left, new DoubleNode(right), operatorPosition);
    }
    public AddNode(double left, double right, int operatorPosition) {
        this(new DoubleNode(left), new DoubleNode(right), operatorPosition);
    }

    @Override
    public char getOperator() {
        return '+';
    }
}

File: shake/compiler/parser/src/commonMain/kotlin/io/github/shakelang/shake/parser/node/expression/AddNode.kt
package io.github.shakelang.shake.parser.node.expression

import io.github.shakelang.shake.parser.node.ValuedNode
import io.github.shakelang.shake.parser.node.factor.DoubleNode
import io.github.shakelang.shake.parser.node.factor.IntegerNode
import io.github.shakelang.parseutils.characters.position.PositionMap

@Suppress("unused")
class AddNode(map: PositionMap, left: ValuedNode, right: ValuedNode, operatorPosition: Int) :
    ExpressionNode(map, left, right, operatorPosition) {
    constructor(map: PositionMap, left: Int, right: ValuedNode, operatorPosition: Int) : this(
        map,
        IntegerNode(map, left),
        right,
        operatorPosition
    )

    constructor(map: PositionMap, left: ValuedNode, right: Int, operatorPosition: Int) : this(
        map,
        left,
        IntegerNode(map, right),
        operatorPosition
    )

    constructor(map: PositionMap, left: Int, right: Int, operatorPosition: Int) : this(
        map,
        IntegerNode(map, left),
        IntegerNode(map, right),
        operatorPosition
    )

    constructor(map: PositionMap, left: Double, right: ValuedNode, operatorPosition: Int) : this(
        map,
        DoubleNode(map, left),
        right,
        operatorPosition
    )

    constructor(map: PositionMap, left: ValuedNode, right: Double, operatorPosition: Int) : this(
        map,
        left,
        DoubleNode(map, right),
        operatorPosition
    )

    constructor(map: PositionMap, left: Double, right: Double, operatorPosition: Int) : this(
        map,
        DoubleNode(map, left),
        DoubleNode(map, right),
        operatorPosition
    )

    override val operator: Char
        get() = '+'

    override fun toJson(): Map<String, *> = mapOf("name" to "AddNode", "left" to left, "right" to right)
}